File: sim/out_cmd/src/sim_dut.cpp
#include "sim_dut.hpp"

#include <iostream>

#include "tb_out_cmd_task_icd_pkg.h"
#include "tb_out_cmd_cmd_icd_pkg.h"

SimDut::SimDut()
{
  Verilated::traceEverOn(true);
  mDut.trace(&mTrace, 5);
  mTrace.open("waveform.vcd");

  std::cout << "=== SIMULATION BEGIN ===\n";
}

SimDut::~SimDut()
{
  std::cout << "=== SIMULATION END ===\n";
  std::cout << "Tests: " << mTestCount << '\n';
  std::cout << "Fails: " << mFailCount << '\n';
  mDut.final();
  mTrace.close();
}

vluint64_t SimDut::get_fail_count()
{
  return mFailCount;
}


void SimDut::task(vluint32_t out, vluint32_t len, uint32_t cmd_ready_latency)
{
    ++mTestCount;
    
    // compute expected outputs
    const vluint32_t e_resp = expected_resp(out, len, cmd_ready_latency);
    const vluint32_t e_cmd  = expected_cmd(out);

    // strobe command into the DUT
    mDut.len_bytes  = len;
    mDut.out_strobe = out;
    mDut.task_valid = 1;
    wait_cycles(1);
    mDut.task_valid = 0;

    // prep the output stream for potential command and wait
    // for response
    vluint32_t a_cmd = 0xFFFFFFFF;
    mDut.aso_cmd_ready = 0;

    // some for loop abuse to handle the latency counter
    for(uint32_t latency = 0; 0 == mDut.resp_valid; ++latency) {
        
        // delay asserting ready
        if(latency > cmd_ready_latency) {
            mDut.aso_cmd_ready = 1;
        }

        // capture the command data
        if (mDut.aso_cmd_ready && mDut.aso_cmd_valid) {
            a_cmd = mDut.aso_cmd_data;
        }

        wait_cycles(1);
    }

    // de-assert stream strobe
    mDut.aso_cmd_ready = 0;
    wait_cycles(1);

    // collect the actual response
    const vluint32_t a_resp = mDut.resp;
    
    // validate results
    if (e_resp != a_resp) {
        mFailCount++;
        std::cout << "-------------------------------------------------\n";
        std::cout << "Response failed\n";
        std::cout << "out = " << out << " len = " << len << " latency = " << cmd_ready_latency << '\n';
        std::cout << "Expected: " << e_resp << '\n';
        std::cout << "Actual  : " << a_resp << '\n';
        std::cout << "-------------------------------------------------\n";
    }

    if (e_resp == tb_out_cmd_task_icd_pkg::TASK_VALID && e_cmd != a_cmd) {
        mFailCount++;
        std::cout << "-------------------------------------------------\n";
        std::cout << "Command failed\n";
        std::cout << "out = " << out << " len = " << len << " latency = " << cmd_ready_latency << '\n';
        std::cout << "Expected: " << e_cmd << '\n';
        std::cout << "Actual  : " << a_cmd << '\n';
        std::cout << "-------------------------------------------------\n";
    }
}

void SimDut::exeErrorTest()
{
    std::cout << "Execution Error Test\n";
    reset();
    task(0, VALID_LEN, TIMEOUT); // valid task that timeout
}

void SimDut::invalidLenTest()
{
    std::cout << "Invalid Length Test\n";
    reset(); 
    task(0, VALID_LEN+1); // length too large
}

void SimDut::outputTest()
{
    constexpr vluint32_t start = 0;
    constexpr vluint32_t end   = 100000;
    std::cout << "Output Test [" << start << " - " << end << "]\n";
    reset();

    for (vluint32_t out = start; out <= end; ++out) {
        task(out, VALID_LEN); 
    }
}

void SimDut::reset(vluint64_t cycles)
{
  mDut.rst           = 1;
  mDut.task_valid    = 0;
  mDut.len_bytes     = 0;
  mDut.out_strobe    = 0;
  mDut.aso_cmd_ready = 0;
  wait_cycles(2);
  
  mDut.rst = 0;
  wait_cycles(1);
}

void SimDut::wait_cycles(vluint64_t cycles)
{
  const vluint64_t target_count = cycles + mPosEdgeCnt;

  while(target_count != mPosEdgeCnt) {
    tick();
  }
}

constexpr vluint32_t SimDut::expected_resp(vluint32_t out, vluint32_t len, uint32_t latency)
{
    if (len != VALID_LEN) {
        return tb_out_cmd_task_icd_pkg::HEADER_INVALID;
    } else if (out > OUT_MAX) {
        return tb_out_cmd_task_icd_pkg::PAYLOAD_INVALID;
    } else if (latency > TIMEOUT) {
        return tb_out_cmd_task_icd_pkg::EXE_ERROR;
    }
    
    return tb_out_cmd_task_icd_pkg::TASK_VALID;
}

constexpr vluint32_t SimDut::expected_cmd(vluint32_t out)
{
    const uint32_t id = tb_out_cmd_cmd_icd_pkg::CMD_ID_OUT <<
        tb_out_cmd_cmd_icd_pkg::CMD_ID_LSB;

    const uint32_t o = out << tb_out_cmd_cmd_icd_pkg::OUT_CMD_LSB; 
    return id | o;
}

void SimDut::tick()
{ 
  mDut.clk ^= 1;
  mDut.eval();

  if (1 == mDut.clk) {
    ++mPosEdgeCnt;
  }

  mTrace.dump(mSimTime);
  mSimTime++;
}

File: sim/out_cmd/sim_dut.py
import cocotb
import os
import sys

from cocotb.clock import Clock
from cocotb.triggers import RisingEdge

# Import files from the packages directory under the sim folder of the
# project.
cwd  = os.path.abspath(os.path.curdir)
pkgs = os.path.join(cwd, "..", "packages")
sys.path.insert(0, os.path.abspath(pkgs))

import task_icd_pkg

class SimDut():

    # Valid length of an output task
    VALID_LEN = task_icd_pkg.HEADER_BYTES + 4

    # Minimum valid output
    OUT_MIN = 0

    # Maximum valid output
    OUT_MAX = 31

    # Execution timeout cycle count
    __TIMEOUT = 1000

    # Maximum amount of timeout latency before the simulation
    # asserts.
    __MAX_LATENCY = __TIMEOUT * 2

    def __init__(self, dut):
        self.__dut = dut
        cocotb.start_soon(Clock(self.__dut.clk, 2, units="ps").start())

    async def act(self, length, output, *, aso_out_valid_latency = 0, force_timeout = False):

        # Add a timeout amount of time to aso_out_valid latency when forcing a
        # timeout condition. Make sure the value isn't bigger than the
        # MAX_LATENCY value.
        if force_timeout:
            new_latency = aso_out_valid_latency + self.__TIMEOUT + 1
            aso_out_valid_latency = min(self.__MAX_LATENCY - 1, new_latency)

        await self.__strobe_cmd(length, output)

        latency  = 0
        cmd_data = list()

        while not self.__dut.resp_valid.value:

            if latency >= self.__MAX_LATENCY:
                assert 0, "DUT took too long to respond"
            elif latency > aso_out_valid_latency:
                cmd_data.append(self.__sink_cmd_word())

            latency += 1
            await self.wait_cycles()

        # Filter out None values without removing 0's
        cmd_data = [d for d in cmd_data if d is not None]

        return (self.__dut.resp.value, cmd_data)


    async def reset(self, cycles = 1):
        self.__dut.rst.value           = 1
        self.__dut.task_valid.value    = 0
        self.__dut.len_bytes.value     = 0
        self.__dut.out_strobe.value    = 0
        self.__dut.aso_cmd_ready.value = 0
        await self.wait_cycles(cycles)
        self.__dut.rst.value = 0
        await self.wait_cycles(1)


    async def wait_cycles(self, cycles = 1):
        for _ in range(cycles):
            await RisingEdge(self.__dut.clk)


    def __sink_cmd_word(self):
        cmd_data = None

        if self.__dut.aso_cmd_valid.value and self.__dut.aso_cmd_ready.value:
            cmd_data = self.__dut.aso_cmd_data.value
            self.__dut.aso_cmd_ready.value = 0
        else:
            self.__dut.aso_cmd_ready.value = 1

        return cmd_data


    async def __strobe_cmd(self, length, output):
        self.__dut.task_valid.value = 1
        self.__dut.len_bytes.value  = length
        self.__dut.out_strobe.value = output
        await self.wait_cycles()
        self.__dut.task_valid.value = 0